### PR TITLE
keccak-data-rlc: compute data RLC in the second phase

### PR DESCRIPTION
**Pull request by naure**
### Description

Compute the RLC of input data in the second phase with a second phase challenge.

### Issue Link

#1016 Keccak Challenge API "Soundness" Bug

Similar to https://github.com/scroll-tech/zkevm-circuits/pull/308

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
    - KECCAK_ROWS defaults changed.
- [ ] This change requires a documentation update

### Contents

- The intermediate computation is witnessed  in unused cells of data_rlc column at rows where the gates are disabled. The computation is an iteration from the previous to the next value, adding one byte at a time.

- KECCAK_ROWS per round must be at least 9 to fit the computation. This can be optimized to 5 or even less if required.

### Rationale

Avoid the introduction of second phase columns in the cell manager.

### How Has This Been Tested?

    cargo test -p zkevm-circuits --features test -- keccak

### CODE REVIEW

File: zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs [hunk 1]
```diff
@@ -574,10 +574,9 @@ pub(crate) fn keccak<F: Field>(
                 transform::value(&mut cell_manager, &mut region, packed, false, |v| *v, true);
             cell_manager.start_region();
             let mut is_paddings = Vec::new();
-            let mut data_rlcs = Vec::new();
+            let mut data_rlcs = vec![Value::known(F::zero()); get_num_rows_per_round()];
```
Comment (ed255) [inline zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:577]:
```suggestion
            let mut data_rlcs = vec![Value::known(F::zero()); NUM_BYTES_PER_WORD + 1];
```
I think this is the exact length we want for `data_rlcs`.

**Review by ed255 — APPROVED** [review]
LGTM! Thanks for porting this PR from scroll's fork!
I've added a small suggestion, please take a look.

File: zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs (hunk 1 above)
Comment (naure) [reply zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:577]: This array holds a value for each row of the round, including rows those not involved in the RLC computation, to be used here:

```rust
            for row_idx in 0..get_num_rows_per_round() {
                rows.push(KeccakRow {
                    q_enable: row_idx == 0,
                    q_round: row_idx == 0 && round < NUM_ROUNDS,
                    q_absorb: row_idx == 0 && round == NUM_ROUNDS,
                    q_round_last: row_idx == 0 && round == NUM_ROUNDS,
                    q_padding: row_idx == 0 && round < NUM_WORDS_TO_ABSORB,
                    q_padding_last: row_idx == 0 && round == NUM_WORDS_TO_ABSORB - 1,
                    round_cst,
                    is_final: is_final_block && round == NUM_ROUNDS && row_idx == 0,
                    length: round_lengths[round],
                    data_rlc: round_data_rlcs[round][row_idx],
                    hash_rlc,
                    cell_values: regions[round].rows[row_idx].clone(),
                });
            }
```

**Comment by naure** [comment]
Thanks for the review! Let me fix the conflicts now.

File: zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs (hunk 1 above)
Comment (ed255) [reply zkevm-circuits/src/keccak_circuit/keccak_packed_multi.rs:577]: Ah I missed that, sorry; you're right!

**Review by CPerezz — APPROVED** [review]
LGTM!

Thanks for this fix @naure !!

**Comment by ed255** [comment]
@naure we can merge this PR once the conflicts are resolved :)

**Comment by naure** [comment]
This was not a conflict, rather that this code has already been merged. Closing.

**Comment by ed255** [comment]
> This was not a conflict, rather that this code has already been merged. Closing.

oooh, so this was merged via https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1315 I totally missed that PR.

**Pull request closed without merging**